### PR TITLE
[executor] detect fork when syncing

### PR DESCRIPTION
**Pull request by lightmark**
## Motivation

When syncing, fork MAY happen due to various reason including deployment issue, bugs. So we add this code snippet to detect it ASAP.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Y
## Test Plan

CI

### CODE REVIEW

File: execution/executor/src/lib.rs [hunk 1]
```diff
@@ -381,19 +383,48 @@ where
...
-            .transaction_infos()
-            .iter()
-            .skip(num_txns_to_skip as usize)
-            .collect();
+        let (skipped_transaction_infos, transaction_infos) = {
+            let (a, b) = txn_list_with_proof
+                .proof
+                .transaction_infos()
+                .split_at(num_txns_to_skip as usize);
+            (a.to_vec(), b.to_vec())
```
Comment (msmouse) [inline execution/executor/src/lib.rs:399]: no need to "to_vec()", right?

File: execution/executor/src/lib.rs [hunk 2]
```diff
@@ -381,19 +383,48 @@ where
...
+        };
+
+        // verify no fork happens.
+        if !txn_list_is_empty {
```
Comment (davidiw) [inline execution/executor/src/lib.rs:403]: any reason to not use what you wrote at line 386?

File: execution/executor/src/lib.rs [hunk 3]
```diff
@@ -381,19 +383,48 @@ where
...
+                    .map(CryptoHash::hash)
+                    .collect::<Vec<_>>()[..],
+            );
+
+            // The two accumulator root hashes should be identical.
+            ensure!(
+                self.cache.synced_trees().state_id() == accu_from_proof.root_hash(),
+                "Fork happens because the current synced_trees doesn't match the txn list provided."
+            )
+        }
```
Comment (msmouse) [inline execution/executor/src/lib.rs:427]: extract to separate method for readability?

File: execution/executor/src/lib.rs [hunk 4]
```diff
@@ -434,7 +465,7 @@ where
             itertools::zip_eq(transactions, output.transaction_data()),
             transaction_infos.into_iter().enumerate(),
         ) {
-            let generated_txn_info = &TransactionInfo::new(
+            let generated_txn_info = TransactionInfo::new(
```
Comment (davidiw) [inline execution/executor/src/lib.rs:468]: what happened here?

File: execution/executor/src/lib.rs [hunk 5]
```diff
@@ -381,19 +383,48 @@ where
...
+            let (a, b) = txn_list_with_proof
+                .proof
+                .transaction_infos()
+                .split_at(num_txns_to_skip as usize);
+            (a.to_vec(), b.to_vec())
+        };
+
+        // verify no fork happens.
+        if !txn_list_is_empty {
+            let frozen_subtree_roots_from_proof = txn_list_with_proof
```
Comment (msmouse) [inline execution/executor/src/lib.rs:404]: add comment, like

"# Left side of the proof happens to be the frozen subtree roots of the accumulator right before the list of txns are applied."

**Comment by wqfish** [comment]
The functions are getting really long. Some refactor would be nice...

File: execution/executor/src/lib.rs (hunk 2 above)
Comment (lightmark) [reply execution/executor/src/lib.rs:403]: This is exactly what I wrote on line 386...

File: execution/executor/src/lib.rs (hunk 1 above)
Comment (lightmark) [reply execution/executor/src/lib.rs:399]: 我傻了

File: execution/executor/src/lib.rs (hunk 2 above)
Comment (davidiw) [reply execution/executor/src/lib.rs:403]: It is the value you produce on line 386. I don't understand why we have line 386. I am proposing you eliminate the seemingly unnecessary expression evaluation on line 386 and do it here. But perhaps there is some Rust moving going on?

File: execution/executor/src/lib.rs (hunk 3 above)
Comment (lightmark) [reply execution/executor/src/lib.rs:427]: I am a little confused about what the inputs should be if get separated. maybe needs more refactoring.

File: execution/executor/src/lib.rs (hunk 4 above)
Comment (lightmark) [reply execution/executor/src/lib.rs:468]: according to changed between 394-399.

**Comment by lightmark** [comment]
> The functions are getting really long. Some refactor would be nice...

I agree. Let's move this job to another PR when I split the executor.

**Pull request closed without merging**